(Z-Image with CREPA added) ValueError: CREPA enabled but unable to infer transformer hidden size.

Full log:
Loading checkpoint shards: 100%|██████████| 2/2 [00:00<00:00, 15.46it/s]
[RANK 0] 2026-04-28 19:18:26,735 [INFO] x86_64-linux-gnu-gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O2 -Wall -fPIC -c /tmp/tmp1jaerg6b/test.c -o /tmp/tmp1jaerg6b/test.o
[RANK 0] 2026-04-28 19:18:26,762 [INFO] x86_64-linux-gnu-gcc /tmp/tmp1jaerg6b/test.o -laio -o /tmp/tmp1jaerg6b/a.out
[RANK 0] 2026-04-28 19:18:26,778 [INFO] x86_64-linux-gnu-gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O2 -Wall -fPIC -c /tmp/tmptuck2q4j/test.c -o /tmp/tmptuck2q4j/test.o
[RANK 0] 2026-04-28 19:18:26,789 [INFO] x86_64-linux-gnu-gcc /tmp/tmptuck2q4j/test.o -L/usr/local/cuda-13.0 -L/usr/local/cuda-13.0/lib64 -lcufile -o /tmp/tmptuck2q4j/a.out
[RANK 0] 2026-04-28 19:18:26,819 [INFO] x86_64-linux-gnu-gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O2 -Wall -fPIC -c /tmp/tmpzxygtm4t/test.c -o /tmp/tmpzxygtm4t/test.o
[RANK 0] 2026-04-28 19:18:26,830 [INFO] x86_64-linux-gnu-gcc /tmp/tmpzxygtm4t/test.o -laio -o /tmp/tmpzxygtm4t/a.out
[RANK 0] 2026-04-28 19:18:27,332 [INFO] Broadcasting error event for job dc325bf2: Training failed: ValueError
[RANK 0] 2026-04-28 19:18:27,333 [INFO] _update_job_store_status called with job_id=dc325bf2, status=failed
[RANK 0] 2026-04-28 19:18:27,333 [INFO] Scheduling _do_updates task for job dc325bf2
[RANK 0] 2026-04-28 19:18:27,333 [INFO] _do_updates starting for job dc325bf2 (status: failed)
[RANK 0] 2026-04-28 19:18:27,335 [INFO] _do_updates: job dc325bf2 is type JobType.LOCAL, current status running
[RANK 0] 2026-04-28 19:18:27,335 [INFO] _do_updates: releasing GPUs for job dc325bf2
[RANK 0] 2026-04-28 19:18:27,336 [INFO] Released GPUs for job dc325bf2
[RANK 0] 2026-04-28 19:18:27,336 [INFO] _do_updates: calling process_pending_jobs
[RANK 0] 2026-04-28 19:18:27,336 [INFO] process_pending_jobs called - checking for pending jobs
[RANK 0] 2026-04-28 19:18:27,337 [INFO] Found 0 pending local jobs
[RANK 0] 2026-04-28 19:18:27,337 [INFO] No pending jobs were started after job dc325bf2 completed
[RANK 0] 2026-04-28 19:18:27,353 [INFO] Unloading all model components...
[RANK 0] 2026-04-28 19:18:27,682 [INFO] Model components unloaded successfully.
Traceback (most recent call last):
  File "/home/gleb/neuro/simpletuner/SimpleTuner/simpletuner/train.py", line 203, in <module>
    trainer.init_load_base_model()
  File "/home/gleb/neuro/simpletuner/SimpleTuner/simpletuner/helpers/training/trainer.py", line 2227, in init_load_base_model
    self.model.load_model(move_to_device=False)
  File "/home/gleb/neuro/simpletuner/SimpleTuner/simpletuner/helpers/models/common.py", line 2944, in load_model
    self.post_model_load_setup()
  File "/home/gleb/neuro/simpletuner/SimpleTuner/simpletuner/helpers/models/z_image/model.py", line 216, in post_model_load_setup
    super().post_model_load_setup()
  File "/home/gleb/neuro/simpletuner/SimpleTuner/simpletuner/helpers/models/common.py", line 5484, in post_model_load_setup
    self._init_crepa_regularizer()
  File "/home/gleb/neuro/simpletuner/SimpleTuner/simpletuner/helpers/models/common.py", line 5502, in _init_crepa_regularizer
    raise ValueError("CREPA enabled but unable to infer transformer hidden size.")
ValueError: CREPA enabled but unable to infer transformer hidden size.
I0428 19:18:27.975000 174068 helpers/logging.py:425] TorchDynamo attempted to trace the following frames: [
I0428 19:18:27.975000 174068 helpers/logging.py:425] 
I0428 19:18:27.975000 174068 helpers/logging.py:425] ]
I0428 19:18:27.977000 174068 helpers/logging.py:425] TorchDynamo compilation metrics:
I0428 19:18:27.977000 174068 helpers/logging.py:425] Function    Runtimes (s)
I0428 19:18:27.977000 174068 helpers/logging.py:425] ----------  --------------
Traceback (most recent call last):
  File "/home/gleb/neuro/simpletuner/.venv/bin/accelerate", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/gleb/neuro/simpletuner/.venv/lib/python3.12/site-packages/accelerate/commands/accelerate_cli.py", line 50, in main
    args.func(args)
  File "/home/gleb/neuro/simpletuner/.venv/lib/python3.12/site-packages/accelerate/commands/launch.py", line 1281, in launch_command
    simple_launcher(args)
  File "/home/gleb/neuro/simpletuner/.venv/lib/python3.12/site-packages/accelerate/commands/launch.py", line 869, in simple_launcher
    raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
subprocess.CalledProcessError: Command '['/home/gleb/neuro/simpletuner/.venv/bin/python3', '/home/gleb/neuro/simpletuner/SimpleTuner/simpletuner/train.py', '--model_family=z_image', '--model_flavour=base', '--controlnet=false', '--output_dir=/home/gleb/.simpletuner/output/ColdGroundClub-oilpaint', '--logging_dir=logs', '--model_type=lora', '--seed=42', '--resolution=1024', '--resume_from_checkpoint=latest', '--pretrained_vae_model_name_or_path=madebyollin/sdxl-vae-fp16-fix', '--vae_dtype=bf16', '--vae_cache_ondemand=false', '--vae_cache_disable=false', '--aspect_bucket_rounding=2', '--base_model_precision=no_change', '--text_encoder_1_precision=no_change', '--text_encoder_2_precision=no_change', '--text_encoder_3_precision=no_change', '--text_encoder_4_precision=no_change', '--offload_during_startup=false', '--delete_model_after_load=false', '--trust_remote_code=false', '--quantize_via=accelerator', '--wan_force_2_1_time_embedding=false', '--pretrained_transformer_subfolder=transformer', '--pretrained_unet_subfolder=unet', '--base_model_default_dtype=bf16', '--max_grounding_entities=0', '--num_train_epochs=0', '--max_train_steps=2500', '--train_batch_size=2', '--learning_rate=1', '--optimizer=prodigy', '--lr_scheduler=constant_with_warmup', '--gradient_accumulation_steps=1', '--lr_warmup_steps=0', '--checkpoints_total_limit=1000', '--gradient_checkpointing', '--gradient_checkpointing_backend=torch', '--enable_group_offload=false', '--ramtorch=false', '--ramtorch_text_encoder=false', '--ramtorch_vae=false', '--ramtorch_controlnet=false', '--ramtorch_transformer_percent=100', '--ramtorch_text_encoder_percent=100', '--ramtorch_disable_sync_hooks=false', '--ramtorch_disable_extensions', '--group_offload_type=block_level', '--group_offload_blocks_per_group=1', '--group_offload_use_stream=false', '--group_offload_text_encoder=false', '--group_offload_vae=false', '--offload_during_save=false', '--train_text_encoder=false', '--lr_num_cycles=1', '--lr_power=0.8', '--use_soft_min_snr=false', '--use_ema=false', '--ema_device=cpu', '--ema_cpu_only=false', '--ema_update_interval=1', '--ema_foreach_disable=false', '--ema_decay=0.995', '--lora_rank=32', '--lora_type=standard', '--slider_lora_target=false', '--lora_dropout=0.1', '--lora_init_type=default', '--peft_lora_mode=standard', '--singlora_ramp_up_steps=0', '--assistant_lora_strength=1.0', '--assistant_lora_inference_strength=0.0', '--disable_assistant_lora=false', '--lora_format=comfyui', '--lycoris_config=/home/gleb/.simpletuner/config/lycoris_config.json', '--use_dora=false', '--resolution_type=pixel_area', '--data_backend_config=/home/gleb/.simpletuner/config/ColdGroundClub-oilpaint/multidatabackend.json', '--caption_strategy=textfile', '--conditioning_multidataset_sampling=random', '--ignore_missing_files=false', '--vae_cache_scan_behaviour=recreate', '--vae_enable_slicing=false', '--vae_enable_tiling=false', '--vae_enable_patch_conv=false', '--vae_enable_temporal_roll=false', '--vae_batch_size=4', '--caption_dropout_probability=0.1', '--audio_channels=1', '--audio_duration_interval=3.0', '--audio_truncation_mode=beginning', '--validation_step_interval=100', '--disable_benchmark=false', '--validation_preview=false', '--validation_preview_steps=1', '--num_eval_images=4', '--eval_timesteps=28', '--eval_dataset_pooling=false', '--eval_loss_disable=false', '--evaluation_type=none', '--pretrained_evaluation_model_name_or_path=openai/clip-vit-large-patch14-336', '--validation_guidance=4', '--validation_num_inference_steps=30', '--validation_on_startup=false', '--validation_method=simpletuner-local', '--validation_external_background=false', '--validation_using_datasets=false', '--validation_randomize=false', '--validation_multigpu=batch-parallel', '--validation_disable=false', '--validation_prompt_library=false', '--user_prompt_library=/tmp/simpletuner_prompt_libraries/dc325bf2/user_prompt_library-ColdGroundClub-oilpaint.json', '--validation_stitch_input_location=left', '--validation_disable_unconditional=false', '--validation_guidance_skip_layers_start=0.01', '--validation_guidance_skip_layers_stop=0.2', '--validation_guidance_skip_scale=2.8', '--validation_lycoris_strength=1.0', '--validation_resolution=256', '--validation_seed_source=cpu', '--validation_adapter_strength=1.0', '--validation_adapter_mode=adapter_only', '--i_know_what_i_am_doing=false', '--diff2flow_enabled=false', '--diff2flow_loss=false', '--scheduled_sampling_max_step_offset=0', '--scheduled_sampling_strategy=uniform', '--scheduled_sampling_probability=0.0', '--scheduled_sampling_prob_start=0.0', '--scheduled_sampling_prob_end=0.5', '--scheduled_sampling_ramp_steps=0', '--scheduled_sampling_start_step=0', '--scheduled_sampling_ramp_shape=linear', '--scheduled_sampling_sampler=unipc', '--scheduled_sampling_order=2', '--scheduled_sampling_reflexflow=false', '--scheduled_sampling_reflexflow_alpha=1.0', '--scheduled_sampling_reflexflow_beta1=10.0', '--scheduled_sampling_reflexflow_beta2=1.0', '--flow_sigmoid_scale=1.0', '--flow_use_uniform_schedule=false', '--flow_use_beta_schedule=false', '--flow_beta_schedule_alpha=2.0', '--flow_beta_schedule_beta=2.0', '--flow_schedule_shift=3.0', '--flow_schedule_auto_shift=false', '--flux_guidance_mode=constant', '--flux_guidance_value=1.0', '--flux_guidance_min=0.0', '--flux_guidance_max=4.0', '--t5_padding=unmodified', '--sd3_clip_uncond_behaviour=empty_string', '--mixed_precision=bf16', '--attention_mechanism=diffusers', '--sageattention_usage=inference', '--disable_tf32=false', '--set_grads_to_none=false', '--noise_offset=0.1', '--noise_offset_probability=0.25', '--input_perturbation=0.0', '--input_perturbation_steps=0', '--lr_end=4e-07', '--lr_scale=false', '--lr_scale_sqrt=false', '--ignore_final_epochs=false', '--freeze_encoder_before=12', '--freeze_encoder_after=17', '--freeze_encoder_strategy=after', '--fully_unload_text_encoder=false', '--save_text_encoder=false', '--text_encoder_limit=100', '--prepend_instance_prompt=false', '--only_instance_prompt=false', '--data_aesthetic_score=7.0', '--delete_unwanted_images=false', '--delete_problematic_images=false', '--disable_bucket_pruning=false', '--allow_dataset_oversubscription=false', '--disable_segmented_timestep_sampling=false', '--preserve_data_backend_cache=false', '--override_dataset_config=false', '--cache_dir_text=cache', '--compress_disk_cache', '--aspect_bucket_disable_rebuild=false', '--keep_vae_loaded=false', '--data_backend_sampling=auto-weighting', '--image_processing_batch_size=32', '--write_batch_size=128', '--read_batch_size=25', '--enable_multiprocessing=false', '--fsdp_enable=false', '--fsdp_version=2', '--fsdp_reshard_after_forward', '--fsdp_state_dict_type=SHARDED_STATE_DICT', '--fsdp_cpu_ram_efficient_loading=false', '--fsdp_auto_wrap_policy=TRANSFORMER_BASED_WRAP', '--fsdp_limit_all_gathers', '--fsdp_cpu_offload=false', '--fsdp_activation_checkpointing=false', '--context_parallel_size=1', '--context_parallel_comm_strategy=allgather', '--num_processes=1', '--num_machines=1', '--dynamo_backend=no', '--dynamo_fullgraph=false', '--dynamo_dynamic=false', '--dynamo_use_regional_compilation=false', '--max_workers=32', '--aws_max_pool_connections=128', '--torch_num_threads=8', '--dataloader_prefetch=false', '--dataloader_prefetch_qlen=10', '--aspect_bucket_worker_count=12', '--aspect_bucket_alignment=64', '--metadata_update_interval=3600', '--debug_aspect_buckets=false', '--debug_dataset_loader=false', '--print_filenames=false', '--print_sampler_statistics=false', '--timestep_bias_begin=0', '--timestep_bias_end=1000', '--timestep_bias_multiplier=1.0', '--timestep_bias_portion=0.25', '--training_scheduler_timestep_spacing=trailing', '--inference_scheduler_timestep_spacing=trailing', '--disk_low_action=stop', '--loss_type=l2', '--huber_schedule=snr', '--huber_c=0.1', '--masked_loss_probability=1.0', '--hidream_use_load_balancing_loss=false', '--crepa_enabled', '--crepa_block_index=8', '--crepa_lambda=0.5', '--crepa_adjacent_distance=1', '--crepa_adjacent_tau=1.0', '--crepa_encoder=dinov2_vitg14', '--crepa_encoder_frames_batch_size=-1', '--crepa_use_backbone_features=false', '--crepa_feature_source=encoder', '--crepa_self_flow=false', '--crepa_self_flow_mask_ratio=0.1', '--crepa_encoder_image_size=518', '--crepa_drop_vae_encoder=false', '--crepa_normalize_by_frames', '--crepa_normalize_neighbour_sum=false', '--crepa_spatial_align', '--crepa_use_tae=false', '--crepa_scheduler=constant', '--crepa_warmup_steps=0', '--crepa_decay_steps=0', '--crepa_lambda_end=0.0', '--crepa_power=1.0', '--crepa_cutoff_step=0', '--crepa_similarity_ema_decay=0.99', '--crepa_threshold_mode=permanent', '--twinflow_enabled=false', '--twinflow_target_step_count=1', '--layersync_enabled=false', '--layersync_lambda=0.2', '--urepa_lambda=0.5', '--urepa_manifold_weight=3.0', '--urepa_model=dinov2_vitg14', '--urepa_encoder_image_size=518', '--urepa_use_tae=false', '--urepa_scheduler=constant', '--urepa_warmup_steps=0', '--urepa_decay_steps=0', '--urepa_lambda_end=0.0', '--urepa_power=1.0', '--urepa_cutoff_step=0', '--urepa_similarity_ema_decay=0.99', '--urepa_threshold_mode=permanent', '--adam_beta1=0.9', '--adam_beta2=0.999', '--adam_weight_decay=0.01', '--adam_epsilon=1e-08', '--max_grad_norm=2.0', '--grad_clip_method=value', '--optimizer_offload_gradients=false', '--fuse_optimizer=false', '--optimizer_release_gradients=false', '--push_to_hub=false', '--push_checkpoints_to_hub=false', '--push_to_hub_background=false', '--model_card_private', '--model_card_safe_for_work=false', '--report_to=tensorboard', '--checkpoint_step_interval=250', '--checkpointing_rolling_steps=0', '--checkpointing_use_tempdir=false', '--checkpoints_rolling_total_limit=1', '--tracker_run_name=run-V2', '--tracker_project_name=ColdGroundClub-oilpaint-Crepa', '--tracker_image_layout=gallery', '--enable_watermark=false', '--seed_for_each_device', '--snr_weight=1.0', '--webhook_config=[{"webhook_type": "raw", "callback_url": "http://0.0.0.0:8001/callback", "log_level": "info", "ssl_no_verify": false, "auth_token": "st_yy4ckWxhcWTKhYgZFkZN2d0cJ6Z2GgMc"}]', '--ema_validation=comparison', '--local_rank=-1', '--offset_noise=false', '--quantize_activations=false', '--refiner_training=false', '--refiner_training_invert_schedule=false', '--refiner_training_strength=0.2', '--sdxl_refiner_uses_full_range=false', '--user_prompt_library=/tmp/simpletuner_prompt_libraries/dc325bf2/user_prompt_library-ColdGroundClub-oilpaint.json']' returned non-zero exit status 1.
[RANK 0] 2026-04-28 19:18:30,726 [ERROR] Function error: ValueError: CREPA enabled but unable to infer transformer hidden size.
[RANK 0] 2026-04-28 19:18:30,727 [ERROR] Traceback (most recent call last):
  File "<string>", line 322, in <module>
  File "/home/gleb/neuro/simpletuner/SimpleTuner/simpletuner/helpers/training/trainer.py", line 7115, in run_trainer_job
    accelerate_result = _launch_with_accelerate()
                        ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gleb/neuro/simpletuner/SimpleTuner/simpletuner/helpers/training/trainer.py", line 7111, in _launch_with_accelerate
    raise launch_error
RuntimeError: ValueError: CREPA enabled but unable to infer transformer hidden size.

[RANK 0] 2026-04-28 19:18:30,730 [INFO] Clearing current_job_id=dc325bf2 due to subprocess status event: error
[RANK 0] 2026-04-28 19:18:30,732 [INFO] Released GPUs for job dc325bf2
[RANK 0] 2026-04-28 19:18:30,732 [INFO] process_pending_jobs called - checking for pending jobs
[RANK 0] 2026-04-28 19:18:30,732 [INFO] Found 0 pending local jobs
[RANK 0] 2026-04-28 19:18:30,735 [INFO] Broadcasting error event for job dc325bf2: ValueError: CREPA enabled but unable to infer transformer hidden size.
[RANK 0] 2026-04-28 19:18:30,736 [INFO] _update_job_store_status called with job_id=dc325bf2, status=failed
[RANK 0] 2026-04-28 19:18:30,736 [INFO] _do_updates starting for job dc325bf2 (status: failed)
[RANK 0] 2026-04-28 19:18:30,738 [INFO] _do_updates: job dc325bf2 is type JobType.LOCAL, current status failed
[RANK 0] 2026-04-28 19:18:30,738 [INFO] _do_updates: job dc325bf2 already in terminal state failed, skipping
[RANK 0] 2026-04-28 19:18:30,739 [INFO] Broadcasting error event for job dc325bf2: ValueError: CREPA enabled but unable to infer transformer hidden size.
[RANK 0] 2026-04-28 19:18:30,740 [INFO] _update_job_store_status called with job_id=dc325bf2, status=failed
[RANK 0] 2026-04-28 19:18:30,740 [INFO] _do_updates starting for job dc325bf2 (status: failed)
[RANK 0] 2026-04-28 19:18:30,741 [INFO] _do_updates: job dc325bf2 is type JobType.LOCAL, current status failed
[RANK 0] 2026-04-28 19:18:30,741 [INFO] _do_updates: job dc325bf2 already in terminal state failed, skipping

--- a/simpletuner/helpers/models/common.py
+++ b/simpletuner/helpers/models/common.py
@@ -5646,6 +5646,10 @@ class ImageModelFoundation(PipelineSupportMixin, VaeLatentScalingMixin, ModelFou
         hidden_size = getattr(config, "hidden_size", None)
         if hidden_size is not None:
             return int(hidden_size)
+        # Fallback: dim (Z-Image and related single-stream transformer configs)
+        dim = getattr(config, "dim", None)
+        if dim is not None:
+            return int(dim)
         return None
 
     def _init_urepa_regularizer(self):

--- a/tests/test_z_image_model.py
+++ b/tests/test_z_image_model.py
@@ -107,6 +107,12 @@ class ZImageModelTests(unittest.TestCase):
         torch.testing.assert_close(received_t, torch.full((1, 16), 0.9, dtype=torch.float32))
         self.assertTrue(torch.equal(output["crepa_hidden_states"], torch.full((1, 16, 8), 7.0)))
 
+    def test_infer_crepa_hidden_size_uses_transformer_dim(self):
+        model = self._build_model()
+        model.model.config = types.SimpleNamespace(dim=3840)
+
+        self.assertEqual(model._infer_crepa_hidden_size(), 3840)
+
     def test_prepare_crepa_self_flow_batch_creates_tokenwise_timesteps(self):
         model = self._build_model()
         model.sample_flow_sigmas = mock.MagicMock(

